File: 06/95/m695.cpp
// Solved:
// (M) Max Area of Island
// https://leetcode.com/problems/max-area-of-island/

#include <queue>
#include <vector>

using namespace std;

class Solution
{
public:
	int maxAreaOfIsland(vector<vector<int>>& grid)
	{
		queue<uint16_t> que;
		int	nMax = 0, cRows = grid.size(), cCols = grid[0].size();
		int i, j, curr, ci, cj, n;

		for (i = 0; i < cRows; ++i)
		{
			for (j = 0; j < cCols; ++j)
			{
				if (grid[i][j])
				{
					n = 0;
					que.push((i << 8) + j);
					while (!que.empty())
					{
						curr = que.front(); que.pop();
						ci = curr >> 8;	cj = curr & 0xFF;
						if (grid[ci][cj])
						{
							n++;
							if (ci > 0) que.push(((ci - 1) << 8) + cj);		// Upper
							if (ci < cRows - 1) que.push(((ci + 1) << 8) + cj);// Lower
							if (cj > 0) que.push((ci << 8) + cj - 1);			// Left
							if (cj < cCols - 1) que.push((ci << 8) + cj + 1);	// Right
							grid[ci][cj] = 0;
						}
					}

					if (n > nMax) nMax = n;
				}
			}
		}

		return nMax;
	}
};

File: 06/95/m695.swift
// Solved
// Helper: `Swift` queue using single-linked list

public class SLListNode<T> {
	fileprivate var value: T
	fileprivate var next: SLListNode?

	public init(_ value: T) {
		self.value = value
	}
}

public class SLQueue<T> {
	private var head: SLListNode<T>?
	private var tail: SLListNode<T>?
	
	public func push(_ element: T) {
		if !isEmpty() {
			tail?.next = SLListNode<T>(element)
			tail = tail?.next
		} else {
			tail = SLListNode<T>(element)
			head = tail
		}
	}
	
	public func pop() -> T? {
		if !isEmpty() {
			let ans = head?.value
			head = head?.next
			if head == nil {
				tail = nil
			}
			return ans
		}
		return nil
	}
	
	public func isEmpty() -> Bool {
		return head == nil
	}
}


class Solution {
	func maxAreaOfIsland(_ grid: [[Int]]) -> Int {
		let que = SLQueue<(Int, Int)>()
		let xMax = grid.count - 1, yMax = grid[0].count - 1
		var g = grid
		var maxArea = 0
		
		for x in 0...xMax {
			for y in 0...yMax {
				if g[x][y] == 1 {
					var area = 0
					
					que.push((x, y))
					while !que.isEmpty() {
						if let (cx, cy) = que.pop() {
							if g[cx][cy] == 1 {
								g[cx][cy] = 0
								if cx > 0 && g[cx - 1][cy] == 1 { que.push((cx - 1, cy)) }
								if cy > 0 && g[cx][cy - 1] == 1 { que.push((cx, cy - 1)) }
								if cx < xMax && g[cx + 1][cy] == 1 { que.push((cx + 1, cy)) }
								if cy < yMax && g[cx][cy + 1] == 1 { que.push((cx, cy + 1)) }
								area += 1
							}
						}
					}
					
					if area > maxArea { maxArea = area }
				}
			}
		}
		
		return maxArea
	}
}


var x = Solution()
let grid = [[0,0,1,0,0,0,0,1,0,0,0,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,1,1,0,1,0,0,0,0,0,0,0,0],[0,1,0,0,1,1,0,0,1,0,1,0,0],[0,1,0,0,1,1,0,0,1,1,1,0,0],[0,0,0,0,0,0,0,0,0,0,1,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,0,0,0,0,0,0,1,1,0,0,0,0]]

print(x.maxAreaOfIsland(grid))
